app/Http/Controllers/Auth/LoginController.php

```diff
@@ -36,161 +28,78 @@ class LoginController extends BaseController
+        return $this->verifyLogin($request);
     }
 
-    protected function attemptVatsimAuth()
+    protected function verifyLogin(Request $request)
     {
-        $allowSuspended = true;
-        $allowInactive = true;
-        $token = VatsimSSO::requestToken(route('auth-vatsim-sso'), $allowSuspended, $allowInactive);
-        if ($token) {
-            $key = $token->token->oauth_token;
-            $secret = $token->token->oauth_token_secret;
-            Session::put('credentials.vatsim-sso', compact('key', 'secret'));
-
-            return redirect()->to(VatsimSSO::sendToVatsim());
+        try {
+            $accessToken = $this->provider->getAccessToken('authorization_code', [
+                'code' => $request->input('code'),
+            ]);
+        } catch (IdentityProviderException $e) {
+            return redirect()->route('dashboard')->withError('Something went wrong, please try again.');
         }
-        // Check if there was a CURL error code
-        if (VATSIMSSO::error()['code']) {
-            Log::error('VATSIMSSO was unable to reach CERT. Code:'.VATSIMSSO::error()['code'].' Message:'.VATSIMSSO::error()['message']);
 
-            return redirect()->back()->withErrors(['connection' => "We were unable to contact VATSIM's certification service. Please try again later. If this persists, please contact Web Services."]);
+        $resourceOwner = json_decode(json_encode($this->provider->getResourceOwner($accessToken)->toArray()));
+
+        if (
+            ! $resourceOwner->data ||
+            ! $resourceOwner->data->cid ||
+            ! $resourceOwner->data->personal ||
+            ! $resourceOwner->data->personal->name_first ||
+            ! $resourceOwner->data->personal->name_last ||
+            ! $resourceOwner->data->personal->email ||
+            ! $resourceOwner->data ||
+            ! $resourceOwner->data->vatsim ||
+            ! $resourceOwner->data->oauth->token_valid === 'true'
+        ) {
+            return redirect()->route('dashboard')->withError('You cannot use our services unless you provide the relevant permissions upon login. Please try again.');
         }
-        throw new \Exception('SSO failed: '.VatsimSSO::error()['message']);
-    }
-
-    protected function setVatsimAuth($userId)
-    {
-        Auth::guard('vatsim-sso')->loginUsingId($userId);
-    }
 
-    protected function attemptSecondaryAuth()
-    {
-        $member = Auth::guard('vatsim-sso')->user();
-
-        if ($member->hasPassword()) {
-            return redirect()->route('auth-secondary');
-        }
+        $account = $this->completeLogin($resourceOwner, $accessToken);
 
-        $intended = Session::pull('url.intended', route('site.home'));
+        Auth::guard('vatsim-sso')->loginUsingId($account->id);
 
-        Auth::login(Auth::guard('vatsim-sso')->user(), true);
-
-        return redirect($intended);
+        return SecondaryLoginController::attemptSecondaryAuth($account);
```

Comment:
Would it not be more prudent to redirect them to the route that this controller method corresponds to? To me, this doesn't feel like the best approach redirecting to another class which has the sole purpose of handling HTTP requests. SRP and all.

Reply:
Further to this, the model binding should keep the account model in tact if it is in the correct state. Food for thought.

Reply:
Take a look at commit 938f9a7
It becomes pretty hard to test unless I extract this out to another method - thoughts?

Reply:
NFA